p2panda-rs/src/document/document.rs

```diff
@@ -0,0 +1,474 @@
+    permissions: Option<Vec<Author>>,
+}
+
+impl DocumentBuilder {
+    /// Instantiate a new DocumentBuilder with a collection of operations.
+    pub fn new(mut operations: Vec<OperationWithMeta>) -> Self {
+        operations.sort_by(|a, b| a.operation_id().as_str().cmp(b.operation_id().as_str()));
+        Self {
+            operations,
+            permissions: None,
+        }
+    }
+
+    /// Get all operations for this document.
+    pub fn operations(&self) -> Vec<OperationWithMeta> {
+        self.operations.clone()
+    }
+
+    /// Get an iterator over all operations in this document.
+    pub fn operations_iter(&self) -> std::vec::IntoIter<OperationWithMeta> {
+        self.operations.clone().into_iter()
+    }
+
+    /// Add permissions for this Document.
+    pub fn permissions(mut self, permissions: Vec<Author>) -> Self {
+        self.permissions = Some(permissions);
+        self
+    }
+
+    /// Build the document.
+    pub fn build(self) -> Result<Document, DocumentBuilderError> {
+        // find create message
+
+        let collect_create_operation: Vec<OperationWithMeta> =
+            self.operations_iter().filter(|op| op.is_create()).collect();
+
+        if collect_create_operation.len() > 1 {
+            return Err(DocumentBuilderError::MoreThanOneCreateOperation);
+        } else if collect_create_operation.is_empty() {
+            return Err(DocumentBuilderError::NoCreateOperation);
+        }
+
+        let create_operation = collect_create_operation.get(0).unwrap(); // unwrap as we know there is one item
+
+        // Get the author of this document from the create message
+        let author = create_operation.public_key();
+
+        // Get the document id from the create message
+        let document_id = create_operation.operation_id();
+
+        // Get the document id from the create message
+        let schema_hash = create_operation.schema();
+
+        // Normally we would get the schema string from the DB by it's hash
+        let schema = Schema::new(&schema_hash, DOCUMENT_SCHEMA)?;
+
+        // Instantiate graph and operations map
+        let mut graph = IncrementalTopo::new();
+        let mut operations = BTreeMap::new();
+
+        for op in self.operations() {
+            // Validate each operation against the document schema before continuing.
+            // NB. cddl crate not wasm supported yet.
+
+            // schema.validate_operation_fields(&op.fields().unwrap())?;
+
+            // Insert operation into map
+            operations.insert(op.operation_id().as_str().to_owned(), op.to_owned());
+            // Add node to graph
+            graph.add_node(op.operation_id().as_str().to_string());
+        }
+
+        // Derive graph dependencies from all operations' previous_operations field. Apply to graph handling
+        // errors.
+        // nb. I had some problems capturing the actual errors from IncrementalTopo crate... needs another
```

Comment:
Let's look at it together? What errors do you get?

Reply:
It was a borrow checker error, yep, let's check it when we look through all this together.

Reply:
Not using incremental topo anymore, problem solved.